File: tests/core/matrices/codac2_tests_Matrix.cpp
/** 
 *  Codac tests
 * ----------------------------------------------------------------------------
 *  \date       2024
 *  \author     Simon Rohou
 *  \copyright  Copyright 2024 Codac Team
 *  \license    GNU Lesser General Public License (LGPL)
 */

#include <iostream>
#include <catch2/catch_test_macros.hpp>
#include <codac2_Vector.h>
#include <codac2_Matrix.h>

using namespace std;
using namespace codac2;

TEST_CASE("Matrix")
{
  Matrix x({
  	{ -1, 3  },
  	{ -6, -9 }
  });

  CHECK(abs(x) == Matrix({
  	{ 1, 3 },
  	{ 6, 9 }
  }));

  Matrix y({
    { -1.2, 3.9  },
    { -6.2, -9.0 }
  });

  CHECK(floor(y) == Matrix({
    { -2, 3  },
    { -7, -9 }
  }));

  CHECK(ceil(y) == Matrix({
    { -1, 4  },
    { -6, -9 }
  }));

  CHECK(round(y) == Matrix({
    { -1, 4  },
    { -6, -9 }
  }));
}

TEST_CASE("Matrix zero init")
{
  auto x = Matrix(2,2);
  CHECK(x(0,0) == 0);
  CHECK(x(1,0) == 0);
  CHECK(x(0,1) == 0);
  CHECK(x(1,1) == 0);

  Vector v(2);
  CHECK(v(0) == 0);
  CHECK(v(1) == 0);
}

File: tests/core/matrices/codac2_tests_Matrix.py
#!/usr/bin/env python

#  Codac tests
# ----------------------------------------------------------------------------
#  \date       2024
#  \author     Simon Rohou
#  \copyright  Copyright 2024 Codac Team
#  \license    GNU Lesser General Public License (LGPL)

import unittest
from codac import *

class TestMatrix(unittest.TestCase):

  def test_def_Matrix(self):

    x = Matrix([
      [ 1, 2 ],
      [ 3, 4 ],
    ])

    x[0,1] = 42
    x[1,1] = 42

    self.assertTrue(x == Matrix([
      [ 1, 42 ],
      [ 3, 42 ],
    ]))

    # The following simple call was a problem before
    # (order of def of constructors):
    m = Matrix(x)


  def test_fnc_Matrix(self):

    x = Matrix([
      [ -1, 3  ],
      [ -6, -9 ]
    ])

    self.assertTrue(abs(x) == Matrix([
      [ 1, 3 ],
      [ 6, 9 ]
    ]))

    y = Matrix([
      [ -1.2, 3.9  ],
      [ -6.2, -9.0 ]
    ])

    self.assertTrue(floor(y) == Matrix([
      [ -2, 3  ],
      [ -7, -9 ]
    ]))

    self.assertTrue(ceil(y) == Matrix([
      [ -1, 4  ],
      [ -6, -9 ]
    ]))

    self.assertTrue(round(y) == Matrix([
      [ -1, 4  ],
      [ -6, -9 ]
    ]))

  def test_def_Matrix_zero_init(self):

    x = Matrix(2,2)
    print(x)
    self.assertTrue(x(0,0) == 0)
    self.assertTrue(x(1,0) == 0)
    self.assertTrue(x(0,1) == 0)
    self.assertTrue(x(1,1) == 0)

    v = Vector(2)
    self.assertTrue(v[0] == 0)
    self.assertTrue(v[1] == 0)

if __name__ ==  '__main__':
  unittest.main()